Redesign

JorgeCastilloPrz opened the pull request:
### ⚠️ WIP ⚠️

### :pushpin: References
**Issues:**
* Fixes #13 
* Fixes #2 

### :tophat: What is the goal?

* We want `Kollect` to use **Tagless Final** style so everything needs to be parametric to `F`. The main goal is to mimic Fetch redesign [still under works but close to be done](https://github.com/47deg/fetch/pull/155). That will allow us to make `Kollect` polymorphic and be able to run targeting any data types that can provide instances of the typeclasses `ConcurrentEffect`, `Timer`, `Parallel`, and `ContextShift` (like `IO`).

### 💻 Implementation & caveats

* Redesigned `Kollect`, `DataSource` and `DataSourceCache` to be parametric to F. Also the inner details and models as `KollectRequest`, `KollectQuery`, `KollectResult`,  `KollectStatus` and so on.
* I ported tons of things from the Scala concurrent package and runtime libraries. This is just something temporary, I assume I'll chunk it in the near future (mostly because I hate it :D), but It has been done mostly to avoid blocking. I believe it's not even working since it has some caveats but I rathered leaving that in stand by for now.
* Some typeclasses and instances that should be in arrow have been implemented here and will be eventually ported.
* I'm waiting for some important updates in `Arrow` that will be needed like `Concurrent`, `Parallel` and a bunch more. This PR needs to get into stand by mode for now, @pakoito is helping with some ongoing work in the Arrow side.

### 📱 How to Test

* TBD

pakoito commented:
`+5,266 −371`

Ayyy

JorgeCastilloPrz commented:
:D don't worry much about the diff size, there wasn't much done before and it was not compiling / complete. Just some scaffoldings. It was meant to be redone from scratch.

nomisRev commented on kollect-core/src/main/kotlin/kollect/Kollect.kt:
```diff
@@ -1,175 +1,399 @@
+            })
+
+    fun <I, B> product(MF: Monad<F>, fb: Kind<KollectPartialOf<F>, B>): Kollect<F, Tuple2<A, B>> =
+            Unkollect(MF.binding {
```
Consider using `flatMap` instead since it only binds once. The resulting code will have same indentation without the current cost we pay for `binding`.

JorgeCastilloPrz replied:
Yep makes sense. Done now 👍

nomisRev commented on kollect-core/src/main/kotlin/kollect/NonEmptyList.kt:
```diff
@@ -0,0 +1,33 @@
+ * res1: scala.collection.immutable.List[String] = List(odd, even, odd, even, odd)
+ * }}}
+ */
+fun <A, B> Kind<ForNonEmptyList, A>.collect(pf: PartialFunction<A, B>): List<B> =
```
Scala docs & consider rewriting without `PartialFunction`.

nomisRev commented on kollect-core/src/main/kotlin/kollect/syntax.kt:
```diff
@@ -0,0 +1,7 @@
+
+import arrow.effects.typeclasses.Concurrent
+
+infix fun <F, A> A.fetch(CF: Concurrent<F>): Kollect<F, A> = Kollect.just(CF, this)
```
Should we swap the API here?

`1 fetch IO.concurrent()` is a little weird while

```kotlin
IO.concurrent().run {
  fetch 1
}
```

is still a good looking DSL like API.
```suggestion
infix fun <F, A> Concurrent<F>.fetch(a: A): Kollect<F, A> = Kollect.just(this, a)
```

nomisRev commented on kollect-core/src/main/kotlin/kollect/syntax.kt:
```diff
@@ -0,0 +1,7 @@
+
+infix fun <F, A> A.fetch(CF: Concurrent<F>): Kollect<F, A> = Kollect.just(CF, this)
+
+infix fun <F, B> Throwable.fetch(CF: Concurrent<F>): Kollect<F, B> = Kollect.error(CF, this)
```
```suggestion
infix fun <F, B> Concurrent<F>.fetch(throwable: Throwable): Kollect<F, B> = Kollect.error(this, throwable)
```

JorgeCastilloPrz replied:
Sounds good, applied now 👍

nomisRev commented on kollect-core/src/main/kotlin/kollect/syntax/CombineSyntax.kt:
```diff
@@ -0,0 +1,111 @@
+/**
+ * Combines two requests to the same data source.
+ */
+private fun <I, A, F> combineRequests(MF: Monad<F>, x: BlockedRequest<F>, y: BlockedRequest<F>): BlockedRequest<F> {
```
```suggestion
private fun <I, A, F> combineRequests(MF: Monad<F>, x: BlockedRequest<F>, y: BlockedRequest<F>): BlockedRequest<F> = MF.run {
```

JorgeCastilloPrz replied:
Fixed. Thanks

nomisRev commented on kollect-core/src/main/kotlin/kollect/syntax/CombineSyntax.kt:
```diff
@@ -0,0 +1,111 @@
+        val anotherId = secondOp.id
+        if (aId == anotherId) {
+            val newRequest = KollectOne(aId, ds)
+            val newResult = { r: KollectStatus -> MF.tupled(x.result(r), y.result(r)).void(MF) }
```
```suggestion
            val newResult = { r: KollectStatus -> tupled(x.result(r), y.result(r)).unit() }
```

nomisRev commented on kollect-core/src/main/kotlin/kollect/syntax/CombineSyntax.kt:
```diff
@@ -0,0 +1,111 @@
+                        r.result as Map<*, *>
+                        val xResult = r.result[aId].toOption().map { KollectDone(it) }.getOrElse { KollectMissing }
+                        val yResult = r.result[anotherId].toOption().map { KollectDone(it) }.getOrElse { KollectMissing }
+                        MF.tupled(x.result(xResult), y.result(yResult)).void(MF)
```
```suggestion
                        tupled(x.result(xResult), y.result(yResult)).unit()
```

nomisRev commented on kollect-core/src/main/kotlin/kollect/syntax/CombineSyntax.kt:
```diff
@@ -0,0 +1,111 @@
+                        MF.tupled(x.result(xResult), y.result(yResult)).void(MF)
+                    }
+                    is KollectMissing ->
+                        MF.tupled(x.result(r), y.result(r)).void(MF)
```
```suggestion
                        tupled(x.result(r), y.result(r)).unit()
```

nomisRev commented on kollect-core/src/main/kotlin/kollect/syntax/CombineSyntax.kt:
```diff
@@ -0,0 +1,111 @@
+                is KollectDone<*> -> {
+                    r.result as Map<*, *>
+                    val oneResult = r.result[oneId].toOption().map { KollectDone(it) }.getOrElse { KollectMissing }
+                    MF.tupled(x.result(oneResult), y.result(r)).void(MF)
```
```suggestion
                    tupled(x.result(oneResult), y.result(r)).unit()
```

nomisRev commented on kollect-core/src/main/kotlin/kollect/syntax/CombineSyntax.kt:
```diff
@@ -0,0 +1,111 @@
+                    val oneResult = r.result[oneId].toOption().map { KollectDone(it) }.getOrElse { KollectMissing }
+                    MF.tupled(x.result(oneResult), y.result(r)).void(MF)
+                }
+                is KollectMissing -> MF.tupled(x.result(r), y.result(r)).void(MF)
```
```suggestion
                is KollectMissing -> tupled(x.result(r), y.result(r)).unit()
```

nomisRev commented on kollect-core/src/main/kotlin/kollect/syntax/CombineSyntax.kt:
```diff
@@ -0,0 +1,111 @@
+                    val oneResult = r.result[oneId].toOption().map { KollectDone(it) }.getOrElse { KollectMissing }
+                    MF.tupled(x.result(r), y.result(oneResult)).void(MF)
+                }
+                is KollectMissing -> MF.tupled(x.result(r), y.result(r)).void(MF)
```
```suggestion
                is KollectMissing -> tupled(x.result(r), y.result(r)).unit()
```

nomisRev commented on kollect-core/src/main/kotlin/kollect/syntax/CombineSyntax.kt:
```diff
@@ -0,0 +1,111 @@
+        val ds = firstOp.ds
+
+        val newRequest = Batch(combineIdentities(firstOp, secondOp), ds)
+        val newResult = { r: KollectStatus -> MF.tupled(x.result(r), y.result(r)).void(MF) }
```
```suggestion
        val newResult = { r: KollectStatus -> tupled(x.result(r), y.result(r)).unit() }
```

nomisRev commented on kollect-core/src/main/kotlin/kollect/syntax/CombineSyntax.kt:
```diff
@@ -0,0 +1,111 @@
+                is KollectDone<*> -> {
+                    r.result as Map<*, *>
+                    val oneResult = r.result[oneId].toOption().map { KollectDone(it) }.getOrElse { KollectMissing }
+                    MF.tupled(x.result(r), y.result(oneResult)).void(MF)
```
```suggestion
                    tupled(x.result(r), y.result(oneResult)).unit()
```

nomisRev commented on kollect-extensions/src/main/kotlin/kollect/extensions/Kollect.kt:
```diff
@@ -0,0 +1,33 @@
+
+// Kollect ops
+@extension
+interface KollectMonad<F, I> : Monad<KollectPartialOf<F>> {
```
This should be split up in seperate typeclasses.

In Scala having an instance for `Monad` is enough because the implicit system can resolve and provide instances for `Functor`, `Applicative`, etc which is transparent but in the Kotlin/Arrow API we have `Applicative<F>` etc all over as explicit params. I think it makes the API more user-friendly if you can call `Kollect.applicative()` etc depending on what you actually need.

nomisRev commented on kollect-extensions/src/main/kotlin/kollect/extensions/Kollect.kt:
```diff
@@ -0,0 +1,33 @@
+
+    override fun <A> just(a: A): Kollect<F, A> = Kollect.Unkollect(MF().just(KollectResult.Done(a)))
+
+    override fun <A, B> Kind<KollectPartialOf<F>, A>.map(f: (A) -> B): Kollect<F, B> =
```
We should avoid using `Kind` when we can provide a more specific (??) type :D 

```suggestion
    override fun <A, B> KollectOf<F, A>.map(f: (A) -> B): Kollect<F, B> =
```

JorgeCastilloPrz replied:
Done 👍

nomisRev commented on kollect-extensions/src/main/kotlin/kollect/extensions/Kollect.kt:
```diff
@@ -0,0 +1,33 @@
+    override fun <A, B> Kind<KollectPartialOf<F>, A>.map(f: (A) -> B): Kollect<F, B> =
+            fix().map(MF(), f)
+
+    override fun <A, B> Kind<KollectPartialOf<F>, A>.product(fb: Kind<KollectPartialOf<F>, B>): Kollect<F, Tuple2<A, B>> =
```
```suggestion
    override fun <A, B> KollectOf<F, A>.product(fb: KollectOf<F, B>): Kollect<F, Tuple2<A, B>> =
```

JorgeCastilloPrz replied:
Done 👍

nomisRev commented on kollect-extensions/src/main/kotlin/kollect/extensions/Kollect.kt:
```diff
@@ -0,0 +1,33 @@
+    override fun <A, B> Kind<KollectPartialOf<F>, A>.product(fb: Kind<KollectPartialOf<F>, B>): Kollect<F, Tuple2<A, B>> =
+            fix().product<I, B>(MF(), fb)
+
+    override fun <A, B> tailRecM(a: A, f: (A) -> Kind<KollectPartialOf<F>, Either<A, B>>): Kollect<F, B> =
```
```suggestion
    override fun <A, B> tailRecM(a: A, f: (A) -> KollectOf<F, Either<A, B>>): Kollect<F, B> =
```

JorgeCastilloPrz replied:
Done 👍

nomisRev commented on kollect-extensions/src/main/kotlin/kollect/extensions/Kollect.kt:
```diff
@@ -0,0 +1,33 @@
+    override fun <A, B> tailRecM(a: A, f: (A) -> Kind<KollectPartialOf<F>, Either<A, B>>): Kollect<F, B> =
+            Kollect.tailRecM(MF(), a, f)
+
+    override fun <A, B> Kind<KollectPartialOf<F>, A>.flatMap(f: (A) -> Kind<KollectPartialOf<F>, B>): Kollect<F, B> = MF().run {
```
```suggestion
    override fun <A, B> KollectOf<F, A>.flatMap(f: (A) -> KollectOf<F, B>): Kollect<F, B> = MF().run {
```

JorgeCastilloPrz replied:
Done 👍

nomisRev commented on kollect-extensions/src/main/kotlin/kollect/extensions/Kollect.kt:
```diff
@@ -0,0 +1,33 @@
+            Kollect.tailRecM(MF(), a, f)
+
+    override fun <A, B> Kind<KollectPartialOf<F>, A>.flatMap(f: (A) -> Kind<KollectPartialOf<F>, B>): Kollect<F, B> = MF().run {
+        fix().flatMap(MF(), f)
```
```suggestion
        fix().flatMap(this, f)
```

nomisRev commented on kollect-extensions/src/main/kotlin/kollect/extensions/Timer.kt:
```diff
@@ -0,0 +1,109 @@
+            IO.async { conn, cb ->
+                val ref = ForwardCancelable()
+                conn.push(ref.cancel())
+                IO { cb(Unit.right()) }.unsafeRunTimed(duration)
```
This can't work? `unsafeRunTimed` works like `timeOut`, right?

JorgeCastilloPrz replied:
True story, I missunderstood what the method does, thought it was the opposite (delaying the execution). Used an alternative implementation based on the jvm pools for now, as you suggested.

nomisRev replied:
Just to make this complete. I discussed with @JorgeCastilloPrz and we both agreed to leave this like this for now until we move `Timer` to Arrow.

In Arrow we can leave the jvm instance in `arrow-effects-io-extensions` and the n the kotlinx impl can just wrap `delay`.

nomisRev commented on kollect-extensions/src/main/kotlin/kollect/extensions/Clock.kt:
```diff
@@ -0,0 +1,98 @@
+
+    fun clock(): Clock<F>
+
+    override fun realTime(unit: TimeUnit): StateT<F, S, Long> = stateT(MF(), clock().realTime(unit))
```
```suggestion
    override fun realTime(unit: TimeUnit): StateT<F, S, Long> = StateT.liftF(MF(), clock().realTime(unit))
```

nomisRev commented on kollect-extensions/src/main/kotlin/kollect/extensions/Clock.kt:
```diff
@@ -0,0 +1,98 @@
+
+    override fun realTime(unit: TimeUnit): StateT<F, S, Long> = stateT(MF(), clock().realTime(unit))
+
+    override fun monotonic(unit: TimeUnit): StateT<F, S, Long> = stateT(MF(), clock().monotonic(unit))
```
```suggestion
    override fun monotonic(unit: TimeUnit): StateT<F, S, Long> = StateT.liftF(MF(), clock().monotonic(unit))
```

nomisRev commented on kollect-typeclasses/src/main/kotlin/kollect/typeclasses/StateT.kt:
```diff
@@ -0,0 +1,9 @@
+import arrow.data.StateT
+import arrow.typeclasses.Applicative
+
+fun <F, S, A> stateT(AF: Applicative<F>, fa: Kind<F, A>): StateT<F, S, A> =
```
This file can be removed because it can be replaced by `StateT.liftF`

JorgeCastilloPrz replied:
Done 👍

JorgeCastilloPrz commented on kollect-core/src/main/kotlin/kollect/Cache.kt:
```diff
@@ -1,45 +1,63 @@
+interface DataSourceCache<F> {
 
-    fun <A> get(k: DataSourceIdentity): Option<A>
+    fun <I : Any, A : Any> lookup(i: I, ds: DataSource<I, A>): Kind<F, Option<A>>
```
🤔 to me it has no sense to allow null ids (`I`) or null cached values (`A`), but pls let me know if you think I'm missing something.

JorgeCastilloPrz replied:
Let's see what @purrgrammer has to say about that :)

purrgrammer replied:
I'm assuming you'd use `Option` to represent nullability, even though it is possible for `null` to be an `I` or an `A` at run time. If you can restrict the `Any` constraint more to avoid nullability I'd go ahead and do it :+1:

nomisRev replied:
@JorgeCastilloPrz the `Any` constraint isn't there in Fetch but lets leave this in.

@raulraja @pakoito do we have a policy for this in Arrow to restrict nullability in generics?

purrgrammer replied:
I'm assuming you'd use `Option[I]` for allowing `None` values in `I`, even though `null` is a valid value for `I` at run time. If you can make a bigger constraint than `Any` to avoid `null` at run time I'd go for it :+1:

JorgeCastilloPrz replied:
Actually that makes a lot of sense now. But I'm not sure if we'd want to "pollute" the whole API contract with Option values just because they wanna use nullable indexes or value types.

JorgeCastilloPrz replied:
There's some implicit problem when I do that. I've been trying and it has some implications.

Any implementation of the`DataSourceCache` contract needs to type their value as a type `A` on their state (that means they require `A` in the class declaration). One example is `InMemoryCache`. 

When we make `I` and `A` not inherit from `Any`, we'd need to move at least `A` to the class level (`DataSourceCache<F, A>`) which could actually make sense. But then I'd personally also pull `I` up, so you have all the constraints in the class type `DataSourceCache<F, I, A>`.

That might sound fine, but that enforces to pass `I` and `A` all the way from the call sites, and you end up having client code on call sites like:

```kotlin
val io = Kollect.runEnv<ForIO, Int, Tuple2<ListK<Int>, ListK<Int>>>(
    IO.concurrent(), 
    IO.timer(EmptyCoroutineContext), 
    kollect(IO.concurrent()))

val res = io.fix().unsafeRunSync()
```

That's because the call site is not able to infer `I` by itself. Any ideas?

purrgrammer replied:
`DataSourceCache[F, I, A]` wouldn't make sense since it would allow us only to store one type of identity-value pairs. `I` and `A` are parameters to the `insert` and `lookup` functions so the cache can be heterogeneous. See [here](https://github.com/47deg/fetch/blob/master/examples/src/test/scala/JedisExample.scala#L154-L185) for a Redis-backed implementation of `DataSourceCache` that can store heterogenous data types.

Regarding Fetch using [`DataSourceId` and other classes](https://github.com/47deg/fetch/blob/06556e7904c0d0533891a40f5f026b0220bfe5c5/shared/src/main/scala/cache.scala#L26) for wrapping cached values: I used it because someone suggested using [value classes](https://docs.scala-lang.org/overviews/core/value-classes.html) for wrapping data source identities and values, but note that those types don't exist at run time.

JorgeCastilloPrz replied:
We can do the same with inline classes, which I actually removed for simplicity, but that would not save us from the `Any` stuff. If you hadn't those classes in Scala you'd have typed your state on the `InMemoryCache` as  `Map[(String, Any), Any]` (given current Fetch constraints on those types)

purrgrammer replied:
>If you hadn't those classes in Scala you'd have typed your state on the InMemoryCache as Map[(String, Any), Any] (given current Fetch constraints on those types)

Exactly

raulraja replied:
If we need Option to push this, let's use it for the time being and revisit this when we have a more stable API

JorgeCastilloPrz replied:
how would `Option<A>` solve the issue?, you'd still need to provide a type for `A` in the `InMemoryCache `, and that's gonna be `Any`, since we don't have `A` in the type declaration, but until cache methods are called.

JorgeCastilloPrz replied:
Added #15 so we can keep the discussion there about how to iterate on the issue.

JorgeCastilloPrz commented:
Addressed all your comments @nomisRev. Thanks 👍

nomisRev commented on kollect-core/src/main/kotlin/kollect/Cache.kt:
```diff
@@ -1,45 +1,63 @@
 
-}
\ No newline at end of file
+fun <K, V> Map<K, V>.updated(k: K, newVal: V) = this.filterNot { it.key == k } + mapOf(k to newVal)
```
It's defined in `arrow.data.MapK` as an extension on `MapK` but we should refactor the receiver to `Map`.

nomisRev commented on kollect-core/src/main/kotlin/kollect/DataSource.kt:
```diff
@@ -1,73 +1,50 @@
+    fun <F> batch(CF: Concurrent<F>, ids: NonEmptyList<I>): Kind<F, Map<I, A>> = CF.binding {
+        val tuples = KollectExecution.parallel(
+                CF,
+                ids.map { id -> fetch(CF, id).map { v -> id.toT(v) } }
```
Less overhead better readability. I probably should've marked this with _Nit_ :D

JorgeCastilloPrz commented on kollect-core/src/main/kotlin/kollect/Foldable.kt:
```diff
@@ -0,0 +1,7 @@
+import arrow.typeclasses.Foldable
+import arrow.typeclasses.Monad
+
+fun <F, G, A, B> Foldable<F>.foldLeftM(MG: Monad<G>, fa: arrow.Kind<F, A>, z: B, f: (B, A) -> arrow.Kind<G, B>): arrow.Kind<G, B> =
```
Yep why not. Still it's not really needed here, since it just uses `foldM` under the hood which is already available. Removed it.

JorgeCastilloPrz commented on kollect-core/src/main/kotlin/kollect/Functor.kt:
```diff
@@ -0,0 +1,6 @@
+import arrow.Kind
+import arrow.typeclasses.Functor
+
+fun <F, A> Kind<F, A>.void(FF: Functor<F>): Kind<F, Unit> = FF.run { this@void.`as`(Unit) }
```
Done 👍

JorgeCastilloPrz commented:
Addressed all your comments now @nomisRev. Sorry for missing those out, seems like GitHub is hiding some comments inside the diff and not exposing them front page.

JorgeCastilloPrz commented:
@purrgrammer , after 5d54554, I got the extensions segregated but test got broken because I'm not implementing `ap` yet for `Kollect`. Could you help on implementing that? We'd like to make people able to run Kollect also with Kollect applicative and functor, even if they'd still need to provide a monad for the inner type `F` in any case.

purrgrammer commented:
> Now I'm clueless about how to implement Kollect#ap(). Does it make sense to provide Functor or Applicative instances for Kollect? In that case, could you help on implementing ap()? It's okay if it's in Scala.

We've discussed this privately on Slack, `ap` can be formulated in terms of `product`:

`ap(ff, fa) = product(ff, fa).map({ case (f, a) => f(a) })`

JorgeCastilloPrz commented:
Alright, `ap` has been implemented by delegating it to `product`, as you suggested @purrgrammer.
